sdk/appcenter/src/main/java/com/microsoft/appcenter/utils/ApplicationLifecycleListener.java

```diff
@@ -0,0 +1,148 @@
+package com.microsoft.appcenter.utils;
+
+import android.app.Activity;
+import android.app.Application.ActivityLifecycleCallbacks;
+import android.os.Bundle;
+import android.os.Handler;
+import android.support.annotation.NonNull;
+import android.support.annotation.Nullable;
+
+import java.util.ArrayList;
+import java.util.Collection;
+
+/**
+ * Listens to the whole application (if the application is a single process) lifecycle events.
+ * It allows catching of a foregrounded and a backgrounded state of the whole application.
+ * It won't fire any events when there's a configuration change in a foregrounded app or if
+ * user goes through activities inside a single application.
+ */
+public class ApplicationLifecycleListener implements ActivityLifecycleCallbacks {
+
+    /**
+     * The timeout used to determine an actual transition to the background.
+     */
+    private static final long TIMEOUT_MS = 700;
+
+    /**
+     * Counter of onActivityStarted() minus onActivityStopped() events.
+     */
+    private int mStartedCounter = 0;
+
+    /**
+     * Counter of onActivityResumed() minus onActivityPaused() events.
+     */
+    private int mResumedCounter = 0;
+
+    /**
+     * Flag indicating that the last activity is paused.
+     */
+    private boolean mPauseSent = true;
+
+    /**
+     * Flag indicating that the last activity is stopped.
+     */
+    private boolean mStopSent = true;
+
+    /**
+     * Background thread handler.
+     */
+    private Handler mHandler;
+
+    /**
+     * A collection of ApplicationLifecycleCallbacks.
+     */
+    private Collection<ApplicationLifecycleCallbacks> mLifecycleCallbacks = new ArrayList<>();
```

Comment:
Use a CopyOnWriteArraySet for listeners. It's registered in background but the event is foreached in UI thread. Also we always use the Set interface for listeners as it's unique registration (avoid duplicates).